onnxruntime/core/providers/dml/dml_provider_factory.cc

```diff
@@ -92,10 +95,301 @@ bool IsSoftwareAdapter(IDXGIAdapter1* adapter) {
+}
+
+static std::optional<OrtDmlPerformancePreference> ParsePerformancePreference(const ProviderOptions& provider_options)
+{
+  static const std::string PerformancePreference = "performance_preference";
+  static const std::string Default = "default";
+  static const std::string HighPerformance = "high_performance";
+  static const std::string MinimumPower = "minimum_power";
+
+  auto preference_it = provider_options.find(PerformancePreference);
+  if (preference_it != provider_options.end()) {
+    if (preference_it->second == Default) {
+      return OrtDmlPerformancePreference::Default;
+    }
+
+    if (preference_it->second == HighPerformance) {
+      return OrtDmlPerformancePreference::HighPerformance;
+    }
+
+    if (preference_it->second == MinimumPower) {
+      return OrtDmlPerformancePreference::MinimumPower;
+    }
+
+    ORT_THROW("Invalid PerformancePreference provided for DirectML EP device selection.");
+  }
+
+  return {};
+}
+
+static std::optional<OrtDmlDeviceFilter> ParseFilter(const ProviderOptions& provider_options)
+{
+  static const std::string Filter = "filter";
+  static const std::string Any = "any";
+  static const std::string Gpu = "gpu";
+  static const std::string Npu = "npu";
+
+  auto preference_it = provider_options.find(Filter);
+  if (preference_it != provider_options.end()) {
+    if (preference_it->second == Any) {
+      return OrtDmlDeviceFilter::Any;
+    }
+
+    if (preference_it->second == Gpu) {
+      return OrtDmlDeviceFilter::Gpu;
+    }
+
+    if (preference_it->second == Npu) {
+      return OrtDmlDeviceFilter::Npu;
+    }
+
+    ORT_THROW("Invalid Filter provided for DirectML EP device selection.");
+  }
+
+  return {};
+}
+
+static std::optional<int> ParseDeviceId(const ProviderOptions& provider_options)
+{
+  static const std::string DeviceId = "device_id";
+
+  auto preference_it = provider_options.find(DeviceId);
+  if (preference_it != provider_options.end()) {
+     if (!preference_it->second.empty()) {
+       return std::stoi(preference_it->second);
+     }
+  }
+
+  return {};
+}
+
+std::shared_ptr<IExecutionProviderFactory> DMLProviderFactoryCreator::CreateFromProviderOptions(
+  const ProviderOptions& provider_options) {
+  auto device_id = ParseDeviceId(provider_options);
+  if (device_id.has_value())
+  {
+    return onnxruntime::DMLProviderFactoryCreator::Create(device_id.value());
+  }
+
+  auto preference = ParsePerformancePreference(provider_options);
+  auto filter = ParseFilter(provider_options);
+
+  // If no preference/filters are specified then create with default preference/filters.
+  if (!preference.has_value() && !filter.has_value())
+  {
+    return onnxruntime::DMLProviderFactoryCreator::CreateFromOptions(nullptr);
+  }
+
+  if (!preference.has_value())
+  {
+    preference = OrtDmlPerformancePreference::Default;
+  }
+
+  if (!filter.has_value())
+  {
+    filter = OrtDmlDeviceFilter::Gpu;
+  }
```

Comment:
Internal consistency within the file, other ifs in this file don't have the opening brace on its own line.